prefect_aws/s3.py

```diff
@@ -576,3 +631,437 @@ def _write_sync(self, key: str, data: bytes) -> None:
+            from prefect_aws.s3 import S3Bucket
+
+            s3_bucket = S3Bucket.load("my-bucket")
+            with BytesIO() as buf:
+                s3_bucket.download_object_to_file_object("my_folder/notes.txt", buf)
+            ```
+
+            Download my_folder/notes.txt object to a BufferedWriter.
+            ```python
+            from prefect_aws.s3 import S3Bucket
+
+            s3_bucket = S3Bucket.load("my-bucket")
+            with open("notes.txt", "wb") as f:
+                s3_bucket.download_object_to_file_object("my_folder/notes.txt", f)
+            ```
+        """
+        client = self.aws_credentials.get_s3_client()
+        bucket_path = self._join_bucket_folder(from_path)
+
+        self.logger.debug(
+            f"Preparing to download object from bucket {self.bucket_name!r} "
+            f"path {bucket_path!r} to file object."
+        )
+        await run_sync_in_worker_thread(
+            client.download_fileobj,
+            Bucket=self.bucket_name,
+            Key=bucket_path,
+            Fileobj=to_file_object,
+            **download_kwargs,
+        )
+        self.logger.info(
+            f"Downloaded object from bucket {self.bucket_name!r} path {bucket_path!r} "
+            f"to file object."
+        )
+        return to_file_object
+
+    @sync_compatible
+    async def download_folder_to_path(
+        self,
+        from_folder: str,
+        to_folder: Optional[Union[str, Path]] = None,
+        **download_kwargs: Dict[str, Any],
+    ) -> Path:
+        """
+        Downloads objects *within* a folder (excluding the folder itself)
+        from the S3 bucket to a folder.
+
+        Args:
+            from_folder: The path to the folder to download from.
+            to_folder: The path to download the folder to.
+            **download_kwargs: Additional keyword arguments to pass to
+                `Client.download_file`.
+
+        Returns:
+            The absolute path that the folder was downloaded to.
+
+        Examples:
+            Download my_folder to a local folder named my_folder.
+            ```python
+            from prefect_aws.s3 import S3Bucket
+
+            s3_bucket = S3Bucket.load("my-bucket")
+            s3_bucket.download_folder_to_path("my_folder", "my_folder")
+            ```
+        """
+        if to_folder is None:
+            to_folder = ""
+        to_folder = Path(to_folder).absolute()
+
+        client = self.aws_credentials.get_s3_client()
+        objects = await self.list_objects(folder=from_folder)
+
+        # do not call self._join_bucket_folder for filter
+        # because it's built-in to that method already!
+        # however, we still need to do it because we're using relative_to
+        bucket_folder = self._join_bucket_folder(from_folder)
+
+        async_coros = []
+        for object in objects:
+            bucket_path = Path(object["Key"]).relative_to(bucket_folder)
+            if bucket_path.is_dir():
```

Comment:
does this mean that any directories (which may themselves contain files) within `from_folder` will not be downloaded?

Reply:
Sub-directories will be downloaded; it's only if it's literally a directory.

I can add a comment.